spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/NDVSketchGenerator.java

```diff
@@ -0,0 +1,120 @@
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.functions;
+import scala.Tuple2;
+
+public class NDVSketchGenerator {
+
+  private NDVSketchGenerator() {}
+
+  public static final String APACHE_DATASKETCHES_THETA_V1_NDV_PROPERTY = "ndv";
+
+  static List<Blob> generateNDVSketchesAndBlobs(
+      SparkSession spark, Table table, long snapshotId, Set<String> columnsToBeAnalyzed) {
+    Map<Integer, ThetaSketchJavaSerializable> columnToSketchMap =
+        computeNDVSketches(spark, table, snapshotId, columnsToBeAnalyzed);
+    return generateBlobs(table, columnsToBeAnalyzed, columnToSketchMap, snapshotId);
+  }
+
+  private static List<Blob> generateBlobs(
+      Table table,
+      Set<String> columns,
+      Map<Integer, ThetaSketchJavaSerializable> sketchMap,
+      long snapshotId) {
+    return columns.stream()
+        .map(
+            columnName -> {
+              Types.NestedField field = table.schema().findField(columnName);
+              Sketch sketch = sketchMap.get(field.fieldId()).getSketch();
+              long ndv = (long) sketch.getEstimate();
+              return new Blob(
+                  StandardBlobTypes.APACHE_DATASKETCHES_THETA_V1,
+                  ImmutableList.of(field.fieldId()),
+                  snapshotId,
+                  table.snapshot(snapshotId).sequenceNumber(),
+                  ByteBuffer.wrap(sketch.toByteArray()),
+                  null,
+                  ImmutableMap.of(APACHE_DATASKETCHES_THETA_V1_NDV_PROPERTY, String.valueOf(ndv)));
+            })
+        .collect(Collectors.toList());
+  }
+
+  private static Map<Integer, ThetaSketchJavaSerializable> computeNDVSketches(
+      SparkSession spark, Table table, long snapshotId, Set<String> columnsToBeAnalyzed) {
+    Map<Integer, ThetaSketchJavaSerializable> sketchMap = Maps.newHashMap();
+    String tableName = table.name();
+    List<String> columns = ImmutableList.copyOf(columnsToBeAnalyzed);
+    Dataset<Row> data =
+        spark
+            .read()
+            .option(SparkReadOptions.SNAPSHOT_ID, snapshotId)
+            .table(tableName)
+            .select(columns.stream().map(functions::col).toArray(Column[]::new));
+    Schema schema = table.schema();
+    List<Types.NestedField> nestedFields =
+        columns.stream().map(schema::findField).collect(Collectors.toList());
+    JavaPairRDD<Integer, ByteBuffer> colNameAndSketchPairRDD =
+        data.javaRDD()
+            .flatMap(
+                row -> {
+                  List<Tuple2<Integer, ByteBuffer>> columnsList =
+                      Lists.newArrayListWithExpectedSize(columns.size());
+                  for (int i = 0; i < row.size(); i++) {
+                    columnsList.add(
+                        new Tuple2<>(
+                            nestedFields.get(i).fieldId(),
+                            Conversions.toByteBuffer(nestedFields.get(i).type(), row.get(i))));
```

Comment:
This conversion doesn't work for Date fields.

This gives the following error -

24/07/18 14:52:56 ERROR Executor: Exception in task 0.0 in stage 1.0 (TID 1)
java.lang.ClassCastException: class java.sql.Date cannot be cast to class java.lang.Integer (java.sql.Date is in module java.sql of loader 'platform'; java.lang.Integer is in module java.base of loader 'bootstrap')
	at org.apache.iceberg.types.Conversions.toByteBuffer(Conversions.java:96)
	at org.apache.iceberg.types.Conversions.toByteBuffer(Conversions.java:83)
	at org.apache.iceberg.spark.actions.NDVSketchGenerator.lambda$computeNDVSketches$8f77ac45$1(NDVSketchGenerator.java:106)
	
I was trying out with the test case "testComputeTableStatsAction" in class "TestComputeTableStatsAction".
The ddl I used was "CREATE TABLE %s (id int, data DATE) USING iceberg"

The same error is seen when we run this on any table having date field.

Reply:
Thanks @jeesou . I believe this is a potential bug in how date type is handled in `org.apache.iceberg.types.Conversions#toByteBuffer(org.apache.iceberg.types.Type.TypeID, java.lang.Object)`
```
      case DATE:
        return ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(0, (int) value);
```
We seem to be casting the date to int which is causing the exception.
I think we could convert the date to its epoch long and store in the byte buffer, and like wise handle it while deserializing in `fromByteBuffer'.
@rdblue @RussellSpitzer thoughts?

Reply:
I'm not sure about your assertion there. java.sql.Date shouldn't get to Conversions.toByteBuffer, that's exclusively for Iceberg's internal type representations. Within Iceberg Date type is an Int so that code is correct.

Reply:
Ah sorry reading the cdoe I see the mistake here, the row object being bassed into Conversions.toByteBuffer is a Spark Row and not an Iceberg one. We would need to convert those types from Spark to Iceberg compatible representations (for all types)

Reply:
Thanks for the clarification @RussellSpitzer.